editor/src/components/editor/store/editor-state.ts

```diff
@@ -2076,9 +2084,158 @@ export const defaultElementWarnings: ElementWarnings = {
   dynamicSceneChildWidthHeightPercentage: false,
 }
 
+export interface RegularNavigatorEntry {
+  type: 'REGULAR'
+  elementPath: ElementPath
+}
+
+export function regularNavigatorEntry(elementPath: ElementPath): RegularNavigatorEntry {
+  return {
+    type: 'REGULAR',
+    elementPath: elementPath,
+  }
+}
+
+export function regularNavigatorEntriesEqual(
+  first: RegularNavigatorEntry,
+  second: RegularNavigatorEntry,
+): boolean {
+  return EP.pathsEqual(first.elementPath, second.elementPath)
+}
+
+export interface ConditionalClauseNavigatorEntry {
+  type: 'CONDITIONAL_CLAUSE'
+  elementPath: ElementPath
+  clause: ThenOrElse
+}
+
+export function conditionalClauseNavigatorEntry(
+  elementPath: ElementPath,
+  clause: ThenOrElse,
+): ConditionalClauseNavigatorEntry {
+  return {
+    type: 'CONDITIONAL_CLAUSE',
+    elementPath: elementPath,
+    clause: clause,
+  }
+}
+
+export function conditionalClauseNavigatorEntriesEqual(
+  first: ConditionalClauseNavigatorEntry,
+  second: ConditionalClauseNavigatorEntry,
+): boolean {
+  return EP.pathsEqual(first.elementPath, second.elementPath) && first.clause === second.clause
+}
+
+export interface SyntheticNavigatorEntry {
+  type: 'SYNTHETIC'
+  elementPath: ElementPath
+  childOrAttribute: ChildOrAttribute
+}
+
+export function syntheticNavigatorEntry(
+  elementPath: ElementPath,
+  childOrAttribute: ChildOrAttribute,
+): SyntheticNavigatorEntry {
+  return {
+    type: 'SYNTHETIC',
+    elementPath: elementPath,
+    childOrAttribute: childOrAttribute,
+  }
+}
+
+export function syntheticNavigatorEntriesEqual(
+  first: SyntheticNavigatorEntry,
+  second: SyntheticNavigatorEntry,
+): boolean {
+  return SyntheticNavigatorEntryKeepDeepEquality(first, second).areEqual
+}
+
+export type NavigatorEntry =
+  | RegularNavigatorEntry
+  | ConditionalClauseNavigatorEntry
+  | SyntheticNavigatorEntry
+
+export function navigatorEntriesEqual(
+  first: NavigatorEntry | null,
+  second: NavigatorEntry | null,
+): boolean {
+  if (first == null) {
+    if (second == null) {
+      return true
+    } else {
+      return false
+    }
+  } else {
+    if (second == null) {
+      return false
+    } else {
+      if (first.type === 'REGULAR' && second.type === 'REGULAR') {
+        return regularNavigatorEntriesEqual(first, second)
+      } else if (first.type === 'CONDITIONAL_CLAUSE' && second.type === 'CONDITIONAL_CLAUSE') {
+        return conditionalClauseNavigatorEntriesEqual(first, second)
+      } else if (first.type === 'SYNTHETIC' && second.type === 'SYNTHETIC') {
+        return syntheticNavigatorEntriesEqual(first, second)
+      } else {
+        return false
+      }
+    }
+  }
+}
```

Comment:
I think that this would be worth simplifying and made more readable:
```ts
  if (first?.type === 'REGULAR' && second?.type === 'REGULAR') {
    return regularNavigatorEntriesEqual(first, second)
  } else if (first?.type === 'CONDITIONAL_CLAUSE' && second?.type === 'CONDITIONAL_CLAUSE') {
    return conditionalClauseNavigatorEntriesEqual(first, second)
  } else if (first?.type === 'SYNTHETIC' && second?.type === 'SYNTHETIC') {
    return syntheticNavigatorEntriesEqual(first, second)
  } else {
    return false
  }
```
(or a variation without the `else`s).

Reply:
Alternatively you could replace this with 2 functions - one that performs the null checks, and then calls the second if both `first` and `second` are non-null values

Reply:
Addressed in this PR: https://github.com/concrete-utopia/utopia/pull/3399